docs/blockstates/introduction.md

```diff
@@ -26,7 +26,52 @@ The log only has one property: axis. A blockstate always has to be defined for a
 "east=false,north=false,south=false,west=false": { "model": "oak_fence_post" }
 ```
 
+!!! Note
+    The properties of a variant string of a blockstate should be dictionary ordered. For example, `"east=false,north=false,south=false,west=false"` is valid, while `"east=false,south=false,west=false,north=false"` is invalid. 
+
 And that is only one variant of 16. This can quickly lead to very big and verbose blockstate files, and is one of the main problems in Minecraft 1.8. Minecraft 1.9 will introduce a system that allows to get this under control. [Forge's Blockstate Json][forge] allows you to do so in 1.8.
 
+Customize state mappers
+--------------------------------
+
+In general, there are one-to-one relations between blockstates of a block and variant strings. However, in some cases, several properties should be ignored when considering block models, such as the growth of a cactus, and whether a block of leaves is decayable. Besides, some blocks have blockstates related to various blockstate json files, such as dirt blocks (`dirt.json`, `coarse_dirt.json`, and `podzol.json`) and sandstones (`sandstone.json`, `chiseled_sandstone.json`, and `smooth_sandstone.json`). When we want to map blockstates to variant strings manually, we need to customize our own state mappers. 
+
+We register our state mappers on the pre-initialization stage like this. Do not forget that it is client side only: 
+
+```java
+ModelLoader.setCustomStateMapper(yourBlock, yourStateMapper);
+```
+
+As the second parameter, all of the state mappers are implementations of an interface, `IStateMapper`. One of the solutions is a subclass of a pre-defined abstract class, `StateMapperBase`. 
+
+The abstract method to be implemented in `StateMapperBase`, `getModelResourceLocation`, provides a map from `IBlockState`, which represents blockstates, to `ModelResourceLocation`, which provides file names (before '#') and variant strings (after '#'). Method `getPropertyString` should be used to serialize properties. 
+
+In addition, a builder called `StateMap.Builder` is generally used instead of our own implementation. an example of vanilla leaves is below: 
+
+```java
+                        // set the property ("variant") which decides the main part
+                        // of file names ("oak", "spruce", "birch", and "jungle")
```

Comment:
"decides the main part of file names" 

Confusing and doesn't get to the point of what this *actually* does.

Here's how I would say it (maybe move this to a paragraph above)

"Calling `withName` and passing it an `IProperty` will cause the blockstate JSON to be split into multiple JSONs based on the values of that property. For example, note that the single block minecraft:stone has four blockstate JSONs stone.json, andesite.json, diorite.json, and granite.json. This is because `withName` is called, passing the stone's VARIANT property. The resulting JSONs are named according to the property's possible values."